.github/actions/arm64-integration/action.yml

```diff
@@ -0,0 +1,53 @@
+name: ARM64 integration tests
+description: Runs the deep integration tests on self-hosted ARM64 runner
+
+inputs:
+  tag:
+    required: true
+  docker-registry:
+    required: true
+  k3d-version:
+    required: true
+
+runs:
+  using: composite
+  steps:
+    - uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8
+    - uses: actions/setup-go@c4a742cab115ed795e34d4513e2cf7d472deb55f
+      with:
+        go-version: '1.18'
+    - uses: docker/setup-buildx-action@94ab11c41e45d028884a99163086648e898eed25
+    - name: Pull linkerd binary
+      shell: bash
+      run: |
+        bin/docker-pull-binaries "${{ inputs.tag }}"
+        CMD="$PWD/target/release/linkerd2-cli-${{ inputs.tag }}-linux-arm64"
+        echo "CMD=$CMD" >> "$GITHUB_ENV"
+        "$CMD" version --client
+    - uses: extractions/setup-just@95b912dc5d3ed106a72907f2f9b91e76d60bdb76
+    - uses: azure/setup-kubectl@7f7e5ba5ea3e491b92e6e8e5819963f8f3a1f076
+    - name: Setup k3d
+      shell: bash
+      run: |
+        mkdir -p $PWD/target/bin
+        PATH=$PATH:$PWD/target/bin
+        echo "PATH=$PATH" >> "$GITHUB_ENV"
+        bin/scurl -v "https://raw.githubusercontent.com/k3d-io/k3d/${{ inputs.k3d-version }}/install.sh" \
+          | USE_SUDO=false K3D_INSTALL_DIR=$PWD/target/bin bash
+    - name: Cluster setup
+      shell: bash
+      run: |
+       CLUSTER_NAME=$(printf "${{ github.ref_name }}" | tr -c [:alnum:] -)
+       echo "CLUSTER_NAME=$CLUSTER_NAME" >> "$GITHUB_ENV"
+       just k3d-name=$CLUSTER_NAME k3d-create
+       just k3d-name=$CLUSTER_NAME k3d-use
```

Comment:
We create the cluster and then in the next step we run the integration tests. At this point in the CI, when the arm tests run, we won't have the images pushed to our registry though, right?

We might have to load the images after `k3d-use`.  Luckily, I think we could merge these two lines, and load the images, by calling into a different recipe altogether.

In the justfile, we set the tag to the output of `bin/tag` -- for a stable release, this should accurately set the stable tag (?)

https://github.com/linkerd/linkerd2/blob/de73ffaa0bc95db83c7b66d7a19b315f48be7a14/justfile#L263

In the `linkerd-load` recipe, we use the tag to load the images:
https://github.com/linkerd/linkerd2/blob/de73ffaa0bc95db83c7b66d7a19b315f48be7a14/justfile#L328-L333

`_linkerd-images` runs first, if we can access the images on the host with `docker ls` (and we should afaik since we do something similar for our normal integration tests, if not we might have to do `docker load` first with tarballs) then all should be well. `_k3d-init` will run second and will create our cluster before the images are loaded. Bonus is, it will also merge the config :)

https://github.com/linkerd/linkerd2/blob/de73ffaa0bc95db83c7b66d7a19b315f48be7a14/justfile#L218-L232

So we could just change this to:
```suggestion
       just k3d-name=$CLUSTER_NAME linkerd-load
```

Or perhaps if we want to be more explicit, create it first then call `linkerd-load`. Do you think this makes sense? Apologies in advance if we are loading the images and I've missed it.

Reply:
Thanks for the detailed write-up! The idea is to have this run in the `release.yml` workflow, replacing the current `arm64_integration_tests` job which depends on `docker_build`, so the images are already published at that point. And IMO this test should exercise having the cluster itself pulling the published images. The other non-arm job `integration_tests` does preload the images, but I think that was added for performance, but the ARM host is pretty swift so it shouldn't need that.

Reply:
Ok, makes sense 👍🏻